File: 2019/ruby/lib/day04.rb
class Day04
  def self.go
    puts "Part1: #{part1}"
    puts "Part2: #{part2}"
  end

  def self.part1
    validate_input_with(:valid?).length
  end

  def self.part2
    validate_input_with(:really_valid?).length
  end

  def self.validate_input_with(validator)
    input_range.each_with_object([]) do |input, valid_passwords|
      valid_passwords << input if Password.new(input.to_s).send(validator)
    end
  end

  def self.input_range
    @input_range ||= Range.new(
      *File.read('../inputs/day04-input.txt')
           .chomp
           .split("-")
           .map(&:to_i)
    )
  end
end

class Password
  def initialize(password)
    @password = password
  end

  def valid?
    return false unless correct_length?
    [ digits_adjacent?,
      ever_increasing?,
    ].all?
  end

  def really_valid?
    # don't bother further evaluation of the password if it's the wrong length
    return false unless correct_length?

    # this chunk_while takes care of finding adjacent digits
    # > "122444".chars.chunk_while do |current_digit, next_digit|
    # *   current_digit == next_digit
    # * end.to_a
    # => [["1"], ["2", "2"], ["444"]]
    @password.chars.chunk_while do |current_digit, next_digit|

      # early return false from the entire method if numbers decrease
      return false if current_digit > next_digit

      # the boolean criteria by which elements are "chunked"
      # put 'em all together if adjacent digits match
      current_digit == next_digit

    # it's a special kind of awful to chain methods on a do/end block
    # but this is the part that checks that any of the collections of
    # adjacent digits are only a pair
    end.to_a.any? { |chunk| chunk.length == 2 }
  end

  def correct_length?
    @password.length == 6
  end

  def digits_adjacent?
    @password.chars.each_cons(2).map do |first, second|
      first == second
    end.any?
  end

  def ever_increasing?
    @password.chars.each_cons(2).map do |first, second|
      first <= second
    end.all?
  end
end

File: 2019/elixir/lib/day04.ex
defmodule Day04 do
  @moduledoc """
  Documentation for Day04.
  """

  def part1 do
    passwords_range()
    |> Enum.filter(&valid?/1)
    |> Enum.count
  end

  @doc """
  Password validation.

  ## Examples

      iex> Day04.valid?("111111")
      true
      iex> Day04.valid?("122345")
      true
      iex> Day04.valid?("111123")
      true
      iex> Day04.valid?("223450")
      false
      iex> Day04.valid?("123789")
      false

  """
  def valid?(password) do
    correct_length?(password) &&
      ever_increasing?(password) &&
      contains_adjacent_digits?(password)
  end

  @doc """
  Password validation in part 2.

  ## Examples

      iex> Day04.really_valid?("111111")
      false
      iex> Day04.really_valid?("112233")
      true
      iex> Day04.really_valid?("111122")
      true
      iex> Day04.really_valid?("123444")
      false
      iex> Day04.really_valid?("123789")
      false

  """
  def really_valid?(password) do
    correct_length?(password) &&
      ever_increasing?(password) &&
      contains_adjacent_digits?(password)
  end

  def passwords_range do
    Aoc2019.get_day_input(__MODULE__)
    |> String.split("-")
    |> Enum.map(&String.to_integer/1)
    |> (fn [a, b] -> a..b end).()
    |> Enum.to_list()
    |> Enum.map(&to_string/1)
  end

  @doc """
  Check password length.

  ## Examples

      iex> Day04.correct_length?("111111")
      true
      iex> Day04.correct_length?("11111")
      false
      iex> Day04.correct_length?("1111111")
      false

  """
  def correct_length?(password) do
    String.length(password) == 6
  end

  @doc """
  all digits are increasing

  ## Examples

      iex> Day04.ever_increasing?("123789")
      true
      iex> Day04.ever_increasing?("122345")
      true
      iex> Day04.ever_increasing?("124454")
      false
      iex> Day04.ever_increasing?("123454")
      false

  """
  def ever_increasing?(password) do
    password
    |> String.graphemes
    |> Enum.map(&String.to_integer/1)
    |> Enum.chunk_every(2, 1, :discard)
    |> Enum.all?(fn [a,b] -> a <= b end)
  end

  @doc """
  contains_adjacent_digits?

  ## Examples

      iex> Day04.contains_adjacent_digits?("123789")
      false
      iex> Day04.contains_adjacent_digits?("122345")
      true
      iex> Day04.contains_adjacent_digits?("124454")
      true
      iex> Day04.contains_adjacent_digits?("123454")
      false

  """
  def contains_adjacent_digits?(password) do
    password
    |> String.graphemes
    |> Enum.chunk_every(2, 1, :discard)
    |> Enum.any?(fn [a,b] -> a == b end)
  end

  @doc """
  contains_adjacent_digits?

  ## Examples

      iex> Day04.contains_a_pair_of_digits?("123789")
      false
      iex> Day04.contains_a_pair_of_digits?("122345")
      true
      iex> Day04.contains_a_pair_of_digits?("124454")
      true
      iex> Day04.contains_a_pair_of_digits?("123454")
      false
      iex> Day04.contains_a_pair_of_digits?("111122")
      true
      iex> Day04.contains_a_pair_of_digits?("123444")
      false

  """
  def contains_a_pair_of_digits?(password) do
    password
    |> String.graphemes
    |> Enum.chunk_while([],
      (fn
        elem, [] -> {:cont, [elem]}
        elem, chunk ->
          if elem == hd(chunk) do
            {:cont, [elem | chunk]}
          else
            {:cont, chunk, [elem]}
          end
       end
      ),
      (fn chunk -> chunk end)
    )
    #|> Enum.any?(fn [a,b] -> a == b end)
  end
end
